REGRESSION(275091@main): Darker output with SVG feComponentTransfer

nullhook opened the pull request:
#### 1c0dff3c8c02031702252129caf96dccdc05cf60
<pre>
REGRESSION(275091@main): Darker output with SVG feComponentTransfer
<a href="https://bugs.webkit.org/show_bug.cgi?id=296776">https://bugs.webkit.org/show_bug.cgi?id=296776</a>
<a href="https://rdar.apple.com/156406312">rdar://156406312</a>

Reviewed by Brent Fulgham and Said Abou-Hallawa.

The regressing commit (275091@main) aimed to fix darker SVG filter
outputs in Cairo by creating SourceGraphic and SourceAlpha to the
operating color space, but omitted the USE(CAIRO) directive,
applying it universally and breaking filter effects in non-Cairo
software appliers.

* LayoutTests/svg/filters/feComponentTransfer-gamma-expected.html: Added.
* LayoutTests/svg/filters/feComponentTransfer-gamma.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectGraph):

Canonical link: <a href="https://commits.webkit.org/298332@main">https://commits.webkit.org/298332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69dc292b69deec12dc3621cdd0f3f55ea9c94ceb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121208 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65733 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/96346812-5a36-4006-bd03-0890da53c365) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116982 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43368 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87461 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ad60399e-6eee-4d59-aa9a-72719b60e651) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118041 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28262 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103331 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67890 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27433 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21451 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64861 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97646 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21568 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124396 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42062 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31458 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96256 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42432 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99521 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96042 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24450 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41246 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19089 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38056 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41937 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47473 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41477 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44796 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43211 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -86,7 +86,11 @@ static std::optional<std::tuple<SVGFilterEffectGraph, FilterEffectGeometryMap>>
+    auto firstPrimitive = childrenOfType<SVGFilterPrimitiveStandardAttributes>(filterElement).first();
+    const auto colorSpace = (firstPrimitive && firstPrimitive->colorInterpolation() == ColorInterpolation::SRGB)
+    ? DestinationColorSpace::SRGB()
+    : DestinationColorSpace::LinearSRGB();
```
Nit: I *think* we would want lines 91 & 92 indented since they are continuations of the prior lines.

brentfulgham reviewed (commented):
Please correct the SaferCPP finding before landing.

shallawa commented on LayoutTests/svg/filters/feComponentTransfer-gamma.html:
```diff
@@ -0,0 +1,34 @@
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta name="fuzzy" content="maxDifference=0-172; totalPixels=0-36515" />
```
This is a large pixel tolerance. It makes it possible that the test may fail and we do not detect the failure.

shallawa commented on LayoutTests/svg/filters/feComponentTransfer-gamma.html:
```diff
@@ -0,0 +1,34 @@
+    </style>
+</head>
+<body>
+    <img id="original-image" class="demo-image" src="resources/room.jpg">
```
Do we have to use image to write this test? Can't we just apply the filter to a rectangle fill with solid color?

shallawa replied:
What is `demo-image`?

shallawa commented on LayoutTests/svg/filters/feComponentTransfer-gamma.html:
```diff
@@ -0,0 +1,34 @@
+    <title>Gamma Filter</title>
+    <style>
+        img {
+            max-width: 500px;
```
Why the style of the image here is different from the expected page:
```
        img {
            width: 320px;
        }
```

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -86,7 +86,18 @@ static std::optional<std::tuple<SVGFilterEffectGraph, FilterEffectGeometryMap>>
+                : DestinationColorSpace::LinearSRGB();
+        }
+
+        return filterElement.colorInterpolation() == ColorInterpolation::SRGB
```
What does this fallback mean? Does it mean we have `<filter></filter>` without any children?

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -86,10 +86,14 @@ static std::optional<std::tuple<SVGFilterEffectGraph, FilterEffectGeometryMap>>
+    const auto colorSpace = filterElement.colorInterpolation() == ColorInterpolation::LinearRGB ? DestinationColorSpace::LinearSRGB() : DestinationColorSpace::SRGB();
+    graph = SVGFilterEffectGraph(SourceGraphic::create(colorSpace), SourceAlpha::create(colorSpace));
+#endif
+
```
Creating `SourceGraphic`, `SourceAlpha` and `SVGFilterEffectGraph` is repeated twice for CAIRO, can we make the code efficient for all platforms?
```
#if USE(CAIRO)
    const auto colorSpace = filterElement.colorInterpolation() == ColorInterpolation::LinearRGB ? DestinationColorSpace::LinearSRGB() : DestinationColorSpace::SRGB();
#else
    const auto colorSpace = DestinationColorSpace::SRGB();
#endif

    SVGFilterEffectGraph graph(SourceGraphic::create(colorSpace), SourceAlpha::create(colorSpace));
    FilterEffectGeometryMap effectGeometryMap;
 ```

nullhook committed:
REGRESSION(275091@main): Darker output with SVG feComponentTransfer
https://bugs.webkit.org/show_bug.cgi?id=296776
rdar://156406312

Reviewed by Brent Fulgham and Said Abou-Hallawa.

The regressing commit (275091@main) aimed to fix darker SVG filter
outputs in Cairo by creating SourceGraphic and SourceAlpha to the
operating color space, but omitted the USE(CAIRO) directive,
applying it universally and breaking filter effects in non-Cairo
software appliers.

* LayoutTests/svg/filters/feComponentTransfer-gamma-expected.html: Added.
* LayoutTests/svg/filters/feComponentTransfer-gamma.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::buildFilterEffectGraph):

Canonical link: https://commits.webkit.org/298332@main
